packages/expect/src/matchers.ts

```diff
@@ -62,32 +70,55 @@ const matchers: MatchersObject = {
       ? () =>
           matcherHint(matcherName, undefined, undefined, options) +
           '\n\n' +
-          `Expected: ${printExpected(expected)}\n` +
-          `Received: ${printReceived(received)}`
+          `Expected: not ${printExpected(expected)}`
       : () => {
           const receivedType = getType(received);
           const expectedType = getType(expected);
-          const suggestToEqual =
-            receivedType === expectedType &&
-            (receivedType === 'object' || expectedType === 'array') &&
-            equals(received, expected, [iterableEquality]);
-          const oneline = isOneline(expected, received);
-          const diffString = diff(expected, received, {expand: this.expand});
+
+          const isShallow =
+            receivedType !== expectedType ||
+            (isPrimitive(expected) &&
+              (expectedType !== 'string' || isOneline(expected, received))) ||
+            expectedType === 'function' ||
+            expectedType === 'regexp' ||
+            (received instanceof Error && expected instanceof Error);
+
+          const deepEqualityName =
+            isShallow || (expectedType !== 'object' && expectedType !== 'array')
```

Comment:
Can you explain why the checks on this line are necessary and why `deepEqualityName` affects `difference` a few lines below? I'm probably missing something, I don't quite understand why the logic needs to be this complex.

Reply:
Renamed `isShallowInequality` and refactored assignments to `deepEqualityName`

Your intuition was correct that `difference` was too complex.

I moved the message below hint and above difference or Expected/Received lines.

Reply:
Cool, now the `deepEqualityName` and `difference` concerns are nicely separated. I've just tried it locally and it looks like you can also remove the `deepEqualityName` conditions? At least it doesn't seem to fail any test.
```diff
-          if (!isShallowInequality) {
-            if (expectedType === 'object' || expectedType === 'array') {
```

Reply:
Yes indeed, thank you for patient critique!

* For `deepEqualityName` the only `if` condition is skip `map` and `set` until we review the current questionable logic
* Separate and rename `hasConciseReport` to make its purpose clear to short-circuit a line diff (while we improve the remaining matchers, we will discover whether it is reusable)
* Added condition and 2 snapshot tests for deep equal and unequal dates